targets/xorg

```diff
@@ -20,28 +20,89 @@ XMETHOD="${XMETHOD:-xorg}"
 # Migration from ambiguous XMETHOD
 rm -f '/etc/crouton/xserverrc-x11'
 
-# On Freon, we need crazy xorg hacks
+
+# Figure out what we need on this system
+
+# Freon?
+freon=''
 if [ ! -f "/sys/class/tty/tty0/active" ]; then
-    compile freon '-ldl' so
+    freon='y'
 fi
 
-ltspackages=''
-# On non-ARM precise, install lts-trusty xorg server to support newer hardware
-# if kernel version != 3.4 (lts-trusty mesa requires version >=3.6)
-if [ "${ARCH#arm}" = "$ARCH" ] && release -eq precise \
-        && ! uname -r | grep -q "^3.4."; then
-    # We still install xorg later to pull in its dependencies
-    ltspackages='-lts-trusty'
-    install --minimal "xserver-xorg$ltspackages" "libgl1-mesa-glx$ltspackages" \
-        "xserver-xorg-input-synaptics$ltspackages"
+# Pull in backported Xorg when possible on precise to support newer hardware
+backport=''
+if release -eq precise; then
+    if [ "${ARCH#arm}" = "$ARCH" ]; then
+        # Note: lts-trusty mesa requires kernel version >=3.6
+        if ! uname -r | grep -q "^3.4."; then
+            backport='-lts-trusty'
+        fi
+    else
+        # ARM only offers quantal backport at the moment
+        backport='-lts-quantal'
+    fi
+fi
+
+# Intel?
+inteldriver=''
+intelhasfbc=''
+pinmesa=''
+intelfbcsupport='y'
+if [ "${ARCH#arm}" = "$ARCH" ]; then
+    inteldriver="xserver-xorg-video-intel$backport"
+
+    # If the system has an Intel FBC-capable video card but has too old a driver
+    # to support the chipset, we'll need to disable hardware acceleration.
+    if grep -q 1 '/sys/module/i915/parameters/i915_enable_fbc' 2>/dev/null; then
+        intelhasfbc=y
+        if release -le wheezy -le kali; then
+            intelfbcsupport=''
+        fi
+    fi
+
+    # On saucy onwards, if kernel version is 3.4, manually pin down old mesa
+    # libraries, as new ones require version >=3.6 (see issue #704).
+    # This is only required on non-Atom Intel chipsets (Atom uses i915 dri driver)
+    if release -ge saucy && uname -r | grep -q "^3.4." &&
+            ! grep -q 0xa0 /sys/class/graphics/fb0/device/device; then
+        pinmesa='y'
+    fi
+fi
+
+# Catalog relevant and irrelevant video drivers
+# Debian sid has modesetting built into xserver-xorg-core
+fbdev="xserver-xorg-video-fbdev$backport"
+modesetting="xserver-xorg-video-modesetting$backport,debian~sid="
+installvideodrivers=''
+removevideodrivers=''
+if [ -n "$freon" ]; then
+    installvideodrivers="$modesetting"
+    if [ -z "$(list_uninstalled "$fbdev")" ]; then
+        removevideodrivers="$fbdev"
+    fi
+else
+    installvideodrivers="$fbdev"
+fi
+if [ -n "$inteldriver" ]; then
+    # If we need SNA (for fbc) but don't have it, don't install intel
+    if [ -z "$intelhasfbc" -o -n "$intelfbcsupport" ]; then
+        installvideodrivers="$installvideodrivers $inteldriver"
+    elif [ -z "$(list_uninstalled "$inteldriver")" ]; then
+        removevideodrivers="$removevideodrivers $inteldriver"
```

Comment:
This only works if the `install` commands below do not accidentally pull in these drivers (we do not have `--minimal` below so this could theoretically happen).

Would it be better to always collect them in `removevideodrivers`, and only remove them if needed later on?